src/optimizer/query_to_operator_transformer.cpp

```diff
@@ -0,0 +1,644 @@
+      LogicalUpdate::Make(target_db_id, target_ns_id, target_table_id, op->GetUpdateClauses()), {});
+
+  if (op->GetUpdateCondition() != nullptr) {
+    std::vector<AnnotatedExpression> predicates;
+    QueryToOperatorTransformer::ExtractPredicates(op->GetUpdateCondition(), &predicates);
+    table_scan = new OperatorExpression(
+        LogicalGet::Make(target_db_id, target_ns_id, target_table_id, predicates, target_table_alias, true), {});
+  } else {
+    table_scan = new OperatorExpression(
+        LogicalGet::Make(target_db_id, target_ns_id, target_table_id, {}, target_table_alias, true), {});
+  }
+  update_expr->PushChild(table_scan);
+
+  output_expr_ = update_expr;
+}
+
+void QueryToOperatorTransformer::Visit(parser::CopyStatement *op, parser::ParseResult *parse_result) {
+  if (op->IsFrom()) {
+    // The copy statement is reading from a file into a table. We construct a
+    // logical external-file get operator as the leaf, and an insert operator
+    // as the root.
+
+    // TODO(Ling): filename? copy statement only has file path
+    auto get_op = new OperatorExpression(
+        LogicalExternalFileGet::Make(op->GetExternalFileFormat(), op->GetFilePath(), op->GetDelimiter(),
+                                     op->GetQuoteChar(), op->GetEscapeChar()),
+        {});
+
+    auto target_table = op->GetCopyTable();
+
+    auto insert_op =
+        new OperatorExpression(LogicalInsertSelect::Make(accessor_->GetDatabaseOid(target_table->GetDatabaseName()),
+                                                         accessor_->GetDefaultNamespace(),
+                                                         accessor_->GetTableOid(target_table->GetTableName())),
+                               {get_op});
+
+    output_expr_ = insert_op;
+
+  } else {
+    if (op->GetSelectStatement() != nullptr) {
+      op->GetSelectStatement()->Accept(this, parse_result);
+    } else {
+      op->GetCopyTable()->Accept(this, parse_result);
+    }
+    auto export_op = new OperatorExpression(
+        LogicalExportExternalFile::Make(op->GetExternalFileFormat(), op->GetFilePath(), op->GetDelimiter(),
+                                        op->GetQuoteChar(), op->GetEscapeChar()),
+        {output_expr_});
+
+    output_expr_ = export_op;
+  }
+}
+
+void QueryToOperatorTransformer::Visit(UNUSED_ATTRIBUTE parser::AnalyzeStatement *op,
+                                       parser::ParseResult *parse_result) {}
+
+void QueryToOperatorTransformer::Visit(parser::ComparisonExpression *expr, parser::ParseResult *parse_result) {
+  auto expr_type = expr->GetExpressionType();
+  if (expr->GetExpressionType() == parser::ExpressionType::COMPARE_IN) {
+    if (GenerateSubqueryTree(expr, 1, parse_result, false)) {
+      // TODO(boweic): Should use IN to preserve the semantic, for now we do not
+      //  have semi-join so use = to transform into inner join
+      // TODO(Ling): now we have semi-join operators. Are we supporting it?
```

Comment:
@amlatyrngom Does the new engine support semi-join?

Regardless of whether we support semi-joins yet, I think it's okay for the binder to handle them correctly. If so, then we can remove these TODOs.